### DISCUSSION TITLE
Command to switch spellcheck language for (human) languages on the fly?

### DISCUSSION THREAD

**Discussion by ttasovac** [streetsidesoftware/vscode-spell-checker#2422] [Q&A] [answered]
Sometimes, it's not advisable to have multiple (human) languages enabled when checking monolingual documents, especially if multiple languages come from the same language family. For instance, Serbian and Russian shouldn't be enabled at the same time because this can lead to false positives, or spelling alternatives being offered from the language that the document is not in.

Manually changing the languages in settings.js is not practical every time we want to make the switch. 

vscode-spell-checker offers handy in-document settings for selecting languages:

`<!-- cSpell:locale ru — >`

but that may not always be ideal —  especially in Markdown files, where we may prefer not to have any such meta comments in the text itself.

Is there a way to write a command that would update `cSpell.language` setting in `settings.js` on the fly? This is what I used to do in Atom (as per https://cschnack.de/blog/2019/atom-spell-lang-swop/), but I have only just started using VSCode and I can't figure out what would correspond to `atom.config.set('spell-check-locales', ['de-DE'])` in the VSCode universe. 

Any thoughts or tips? Many thanks in advance.

**Comment by Jason3S** [comment, marked as answer]
@ttasovac,

A command like that doesn't currently exist for the spell checker. It is a possible enhancement.

## Keybinding to Settings: `cSpell.language`

You could try using this keybinding:

**`keybindings.json`**
```json
    {
        "key": "ctrl+alt+cmd+s",
        "command": "workbench.action.openSettings",
        "args": { "query": "cSpell.language" }
    }
```

<img width="703" alt="image" src="https://user-images.githubusercontent.com/3740137/210319372-0197acc5-d125-45cc-b045-1194767a767c.png">


## CSpell Language Extensions

Each language extension comes with its own command to enable/disable the locale.

It is possible to add keybindings for them:

<img width="712" alt="image" src="https://user-images.githubusercontent.com/3740137/210318903-98087af3-d92b-4c33-b866-3a234ba17f27.png">

<img width="1030" alt="image" src="https://user-images.githubusercontent.com/3740137/210319102-33da3e72-7fbb-4290-b77f-2a7c61c5de00.png">

## Using Overrides

`cSpell.overrides` provides a way to use the file path as a way to select the locale.

<img width="1029" alt="image" src="https://user-images.githubusercontent.com/3740137/210320059-45773a0f-beed-4627-bb1f-d5aa3e8a2e6d.png">

```
    "cSpell.overrides": [
        {
            "filename": "**/i18n/ru/**",
            "language": "ru"
        }
    ]
```

**Comment by ttasovac** [comment]
Thanks a lot, @Jason3S. This is all very useful.

Thanks for accepting my Serbian PR. By the way, do you automatically create VSCode extensions, or should I submit another PR for it?

**Reply by Jason3S** [reply]
@ttasovac,

I appreciate that you made the Serbian dictionaries.

The instructions on how to create an extension are here: [vscode-cspell-dict-extensions/CONTRIBUTING.md](https://github.com/streetsidesoftware/vscode-cspell-dict-extensions/blob/main/CONTRIBUTING.md).

I think it is best to make one extension with both dictionaries. If you create the extension for just one dictionary, I'll help with adding the second and the extra commands.

**Comment by ttasovac** [comment]
> I think it is best to make one extension with both dictionaries. If you create the extension for just one dictionary, I'll help with adding the second and the extra commands.

@Jason3S , I've created and tested it, and both dictionaries work within a single extension. I still need a bit of advice though.

Because I created multiple locales in cspell, they are now all now showing up  VSCode like this:

<img width="718" alt="Снимок экрана 2023-01-07 в 05 33 18" src="https://user-images.githubusercontent.com/5304839/211131102-788499e8-61cb-4828-a1e7-8a77dc4aa083.png">

This looks messy and doesn't allow a user to select either Cyrillic or Latin, you always need to have both enabled.

So, there are two issues I'd like to resolve:

1. I'm not sure how useful are all these country-specific locales. I don't know how many people actually use them, and I think they may end up confusing the users, especially since their presence doesn't effect the spell-checking in any way. Perhaps it would be better if I simplified things in my cSpell dictionaries and if I created a PR to reduce the number of available locales to rs and rs-Cyrl/rs-Latn only? 
2. My ideal setup in VSCode would be to give the user a choice to enable _either_ Latin _or_ Cyrillic _or_ both. But that should be left up to the user. If somebody never uses one of the two scripts, they shouldn't have to have them enabled. I don't see how I can do that if both dictionaries are in one VSCode extension. Is there a way or would I need to create two VSCode extensions: Serbian (Cyrillic) and Serbian (Latin) for this to be possible. And if so, would you be ok with it?


Please let me know.

**Comment by Jason3S** [comment]
Nicely done.

Since there are no differences, then having a single locale `sr` would be sufficient. Having two, implies that there might be differences.

I think it is worth having the three codes+script combos:

- `sr` - for both Serbian - both scripts
- `sr-Latin` - for Latin only
- `sr-Cyrl` - for Cyrillic only

I'll Update the Spell Checker to include the list of scripts.

I think we can go with what you made in a Preview version, then update it when things are fixed.

**Comment by ttasovac** [comment]
Yes, that makes sense: `sr` for both, and then `sr-Cyrl` and `sr-Latn` for individual scripts. 

Just to clarify, I should:

- submit a PR to `cspell-dicts` to use only sr, sr-Cyrl and sr-Latin with the Serbian dictionaries
- submit a PR to `vscode-cspell-dict-extensions` which adds the Serbian extension

And when you say:

> I'll Update the Spell Checker to include the list of scripts.

Does that mean that the users will eventually be able to select either Cyrillic or Latin, or both —  from the UI?

**Reply by Jason3S** [reply]
> Just to clarify, I should:
> 
> * submit a PR to `cspell-dicts` to use only sr, sr-Cyrl and sr-Latin with the Serbian dictionaries
> * submit a PR to `vscode-cspell-dict-extensions` which adds the Serbian extension

yes

> Does that mean that the users will eventually be able to select either Cyrillic or Latin, or both — from the UI?

yes

**Reply by ttasovac** [reply]
👍
